Refactoring code for application printing + fixing #11091 (inconsistencies in parsing/printing notations to partial applications)

herbelin opened the pull request:
**Kind:** bug fix / cleanup

Fixes #11091. A newly added commit now fixes #4690 too.

That is, it makes inheritance of implicit arguments and argument scopes consistent across parsing and printing for notations bound to partial applications. In particular it decides to always inherit implicit arguments except when the notation is to an expression of the form `@foo` (without arguments), in which case implicit arguments are deactivated. (Formerly, no proper so global decision was taken on this kind of matter resulting in a combination of inconsistent heuristics.)

The PR also proposes to inherit argument scopes in notations bound to an expression of the form `@foo`. This is a change of semantics, but I don't find intuitive that `@` would act on argument scopes. 

The PR incidentally also fixes an index bug in the algorithm deciding when an implicit argument cannot be reinferred and has to be made explicit.
 
So, to summarize, in addition to several printing adjustements, the semantic changes at parsing, which I expect to be rare in practice, are:
- In `Notation "..." := (foo ...)` and `Notation "..." := (@foo ...)`, implicit arguments are now inherited, as it was the case for `Notation id := (foo ...)` and `Notation id := (@foo ...)`, for both terms and patterns.
- In `Notation "..." := @foo` and `Notation id := @foo`, argument scopes are now kept (but implicit arguments continue to be not propagated), for both terms and patterns.

- [X] Added / updated test-suite
- [X] changelog
- [x] document these rules in the chapters on notations

ejgallego commented:
Needs rebase; I wonder if the PR could be split as to ease review, minimize future conflicts, and merge quicker.

Zimmi48 commented on doc/changelog/03-notations/11120-master+refactoring-application-printing.rst:
```diff
@@ -0,0 +1,16 @@
+- **Fixed:** Inheritance of implicit arguments across notations made
+  uniform in parsing and printing. To the exception of notations of
+  the form ``Notation "..." := @foo`` and ``Notation id := @foo`` which
```
You could consider using the Sphinx support for things like:

```suggestion
  the form ``Notation @string := @foo`` and ``Notation @ident__1 := @@ident__2`` which
```

Later on you can also use the `@args` token.

herbelin replied:
Strictly speaking the second ident should be a qualid. I'll update.

Zimmi48 commented on doc/changelog/03-notations/11120-master+refactoring-application-printing.rst:
```diff
@@ -0,0 +1,16 @@
+- **Fixed:** Inheritance of implicit arguments across notations made
+  uniform in parsing and printing. To the exception of notations of
```
This first sentence sounds a bit convoluted to me. What about:

>Parsing and printing now handle inherited implicit arguments in notations the same.

? I'm not sure it fits exactly what you meant. In any case, we should try to aim for an active form and avoid the word "made".

herbelin replied:
I made the following proposition: "Parsing and printing consistently handle inheritance of implicit arguments in notations". WDYT?

Zimmi48 commented on doc/changelog/03-notations/11120-master+refactoring-application-printing.rst:
```diff
@@ -0,0 +1,16 @@
+  inhibit implicit arguments, all notations binding a partially
+  applied constant, as e.g. in ``Notation "..." := (foo args)``,
+  or ``Notation "..." := (@foo args)``, or ``Notation id := (foo args)``, or
+  ``Notation id := (@foo args)``, inherits the remaining implicit arguments
```
IINM this is "all notations inherit"

```suggestion
  ``Notation id := (@foo args)``, inherit the remaining implicit arguments
```

Zimmi48 commented on doc/changelog/03-notations/11120-master+refactoring-application-printing.rst:
```diff
@@ -0,0 +1,16 @@
+- **Fixed:** Inheritance of implicit arguments across notations made
+  uniform in parsing and printing. To the exception of notations of
+  the form ``Notation @string := @foo`` and ``Notation @ident := @@qualid`` which
```
Sorry I was a bit sloppy in my examples, you additionally need to replace the double backticks by simple backticks preceded by `:n:`. E.g.:
```suggestion
  the form :n:`Notation @string := @qualid` and :n:`Notation @ident := @@qualid` which
```

herbelin replied:
OK, I made a change, please check. In particular, is my ``n:`@@qualid` `` correct?

Zimmi48 replied:
I think so yes. However, what is not is the `@foo` which is interpreted as a token `foo` which does not exist, which breaks the build. If you absolutely want to keep `foo` here, and not use a relevant token like `qualid`, then we need to escape the `@`. I am not sure how to do this, though.

Zimmi48 commented on doc/changelog/03-notations/11120-master+refactoring-application-printing.rst:
```diff
@@ -0,0 +1,16 @@
+  inhibit implicit arguments, all notations binding a partially
+  applied constant, as e.g. in :n:`Notation @string := (@qualid @args)`,
+  or :n:`Notation @string := (@@qualid @args)`, or :n:`Notation @ident := (@qualid @args)`, or
+  :n:`Notation @ident := (@@qualid @args)`, inherit the remaining implicit arguments
```
I didn't realize before but you can further simplify this list in just two:

```rst
as e.g. in :n:`Notation {| @string | @ident } := (@qualid @args)`
or :n:`Notation {| @string | @ident } := (@@qualid @args)`
```

or even in just one (maybe a bit extreme):

```rst
as e.g. in :n:`Notation {| @string | @ident } := ({? @ } @qualid @args)`
```

herbelin replied:
I prefer not to make the reading to tricky.

herbelin replied:
Apparently, sphinx fails on `@args` (`'token' reference target not found`). I replaced by `{+ @arg }` which seems to be the standard way to express what I meant (a non-empty list of arguments).

maximedenes commented:
> The PR incidentally also fixes an index bug in the algorithm deciding when an implicit argument cannot be reinferred and has to be made explicit.

How hard would it be to extract this fix? It would be good to integrate it, and possibly backport it.

herbelin commented:
> How hard would it be to extract this fix? It would be good to integrate it, and possibly backport it.

It would not be too difficult (it is a bit of arithmetic done in 18dedf66b54f4 to be propagated by continuity in the `exprec` subfunction of `Constrextern.explicitize`), but it is a maybe relatively minor bug in `Defensive` mode. Was it reported?

jfehrle commented on doc/changelog/03-notations/11120-master+refactoring-application-printing.rst:
```diff
@@ -0,0 +1,16 @@
+- **Fixed:** Inheritance of implicit arguments across notations made
+  uniform in parsing and printing. To the exception of notations of
```
```suggestion
  uniform in parsing and printing. With the exception of notations of
```

herbelin commented:
@ejgallego:

> I wonder if the PR could be split as to ease review, minimize future conflicts, and merge quicker.

The four first commits are large but in principle w/o semantics change. They somehow clutter the PR uselessly and I'm going to submit them separately, hoping it helps.

herbelin commented:
I rebased. How can I help reviewing for the 19 remaining commits? These are:
- 6 fixes (among them the anomaly in #11091 and #4690)
- 6 commits about unifying the semantics along the line of #11901
- 6 commits about doc/changelog/overlays/test-suite
- 1 small extra commit about restructuring code.

ejgallego commented:
> I rebased. How can I help reviewing for the 19 remaining commits? These are:

Thanks for the summary @herbelin ; from my part I'm gonna try to process all other notation-based PRs, then address this one.

ejgallego commented:
I had a look, IMHO this looks mostly ready and it addresses and important problem; I have a few comments about the code I'll do shortly.

ejgallego commented on interp/constrextern.ml:
```diff
@@ -736,8 +736,10 @@ let extern_applied_notation n impl f args =
     f.CAst.v
   else
-    let args = adjust_implicit_arguments false (List.length args) 1 args impl in
+  try
```
This `try ... catch` was missing due to the API of adjust_implicit_arguments? Using an exception here seems overkill and prone to errors maybe.

ejgallego commented on interp/notation_ops.mli:
```diff
@@ -69,12 +69,12 @@ val match_notation_constr : bool -> 'a glob_constr_g -> interpretation ->
   'a cases_pattern_g -> interpretation ->
   (('a cases_pattern_g * extended_subscopes) list * ('a cases_pattern_g list * extended_subscopes) list) *
-    (int * 'a cases_pattern_g list)
+    (bool * int * 'a cases_pattern_g list)
```
This is becoming quite hard to read; maybe time to use a record?

ejgallego commented:
> @herbelin have you opened the overlay's PRs?

overlays links still missing

ejgallego commented:
Ups, closed on error sorry. Refman is broken:
```
Warning, treated as error:
 doc/unreleased.rst:53:'token' reference target not found: args
```

herbelin commented:
> have you opened the overlay's PRs?

They were, and, since they were backward-compatible, they were merged at once. The problem is that the overlays were now out of sync. I simply removed the overlay commit.

>  Refman is broken:

I fixed the failure of the changelog.

Finally, I just added a documentation of the rule for inheriting implicit arguments and scopes.

ejgallego commented:
> Finally, I just added a documentation of the rule for inheriting implicit arguments and scopes.

Great, thanks @herbelin ; will merge once the CI is done.
